langchain_benchmarks/rate_limiting.py

```diff
@@ -0,0 +1,107 @@
+from langchain.schema.runnable.utils import Input, Output
+
+
+class RateLimiter:
+    def __init__(
+        self,
+        *,
+        requests_per_second: float = 1,
+        tokens_per_request: int = 1,
+        check_every_n_seconds: float = 0.1,
+    ) -> None:
+        """A rate limiter based on a token bucket.
+
+        This is a rate limiter which works in a threaded environment.
+
+        It works by filling up a bucket with tokens at a given rate. Each
+        request consumes a given number of tokens. If there are not enough
+        tokens in the bucket, the request is blocked until there are enough
+        tokens.
+
+        Args:
+            requests_per_second: The number of tokens to add per second to the bucket.
+                Must be at least 1.
+            tokens_per_request: number of tokens that a request "costs".
+                This is 1 by default.
+            check_every_n_seconds: check whether the tokens are available
+                every this many seconds. Can be a float to represent
+                fractions of a second.
+        """
+        if requests_per_second < 1:
+            raise ValueError("Rate must be at least 1 request per second")
+        self.requests_per_second = requests_per_second
+        # Number of tokens in the bucket.
+        self.available_tokens = 0.0
+        # A lock to ensure that tokens can only be consumed by one thread
+        # at a given time.
+        self._consume_lock = threading.Lock()
+        # tokens per request sets how many tokens
+        self.tokens_per_request = tokens_per_request
+        # The last time we tried to consume tokens.
+        self.last: Optional[time.time] = None
+        self.check_every_n_seconds = check_every_n_seconds
+
+    def consume(self) -> bool:
+        """Consume the given amount of tokens if possible.
+
+        Returns:
+            True means that the tokens were consumed, and the caller can proceed to
+            make the request. A False means that the tokens were not consumed, and
+            the caller should try again later.
+        """
+        with self._consume_lock:
+            now = time.time()
+
+            # initialize on first call to avoid a burst
+            if self.last is None:
+                self.last = now
+
+            elapsed = now - self.last
+
+            if elapsed * self.requests_per_second > 1:
+                self.available_tokens += elapsed * self.requests_per_second
+                self.last = now
+
+            self.available_tokens = min(self.available_tokens, self.requests_per_second)
+
+            if self.available_tokens >= self.tokens_per_request:
+                self.available_tokens -= self.tokens_per_request
+                return True
+
+            return False
+
+    def wait(self) -> None:
+        """Blocking call to wait until the given number of tokens are available."""
+        while not self.consume():
+            time.sleep(self.check_every_n_seconds)
```

Comment:
nit, may be able to use `threading.Condition` with wait / notify if wanting to avoid polling.

Reply:
I don't have much experience with threading primitives. Any suggestions on how to re-write with condition? My understanding is that something will still need to do notification in this case, so at least one thread will need to do polling. is that right or is there a good way to re-write this?

Reply:
Ah yes, that's correct - misread this and thought I saw a thread, but I see that it's being updated inline.  As you mentioned, the way to do that would be to create a background thread that replenishes the tokens and notifies the other threads waiting this that condition. But given we don't have that, would prob leave as is.